File: code/ch04/python/cocomo_basic.py
# §4.6.2 The Cocomo Family of Estimation Models — Basic COCOMO effort and
# schedule for the clinic scheduling app (organic mode: a = 2.4, b = 1.05).
from math import isclose


def effort(kloc, a=2.4, b=1.05):          # Basic COCOMO, organic
  return a * kloc ** b


def schedule(e):
  return 2.5 * e ** 0.38                # calendar months


e20, e40 = effort(20), effort(40)
print(f"20 KLOC: {e20:5.1f} person-months, {schedule(e20):.1f} months")
print(f"40 KLOC: {e40:5.1f} person-months, {schedule(e40):.1f} months")
print(f"doubling factor: {e40 / e20:.2f}")

# Checks against the chapter text.
assert round(e20) == 56                   # "~56 person-months"
assert round(e40, 1) == 115.4             # "≈ 115.4 person-months"
assert round(e40 / e20, 2) == 2.07        # "a factor of 2.07, not 2.0"
assert isclose(e40 / e20, 2 ** 1.05)      # the penalty is set by b alone

File: code/ch04/java/CocomoBasic.java
// §4.6.2 The Cocomo Family of Estimation Models — Basic COCOMO effort and
// schedule for the clinic scheduling app (organic mode: a = 2.4, b = 1.05).
public class CocomoBasic {
  static double effort(double kloc) {          // Basic COCOMO, organic
    return 2.4 * Math.pow(kloc, 1.05);
  }

  static double schedule(double e) {
    return 2.5 * Math.pow(e, 0.38);            // calendar months
  }

  public static void main(String[] args) {
    double e20 = effort(20), e40 = effort(40);
    System.out.printf("20 KLOC: %5.1f person-months, %.1f months%n", e20, schedule(e20));
    System.out.printf("40 KLOC: %5.1f person-months, %.1f months%n", e40, schedule(e40));
    System.out.printf("doubling factor: %.2f%n", e40 / e20);

    // Checks against the chapter text.
    assert Math.round(e20) == 56;                   // "~56 person-months"
    assert Math.round(e40 * 10) == 1154;            // "≈ 115.4 person-months"
    assert Math.round(e40 / e20 * 100) == 207;      // "a factor of 2.07, not 2.0"
    assert Math.abs(e40 / e20 - Math.pow(2, 1.05)) < 1e-9;  // penalty set by b alone
  }
}
